airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/DestinationTaskScopeProvider.kt

```diff
@@ -57,42 +51,60 @@ class DestinationTaskScopeProvider(config: DestinationConfiguration) :
     private val timeoutMs = config.gracefulCancellationTimeoutMs
 
     data class ControlScope(
-        val job: Job,
-        val dispatcher: CoroutineDispatcher,
+        val name: String,
+        val job: CompletableJob,
+        val dispatcher: CoroutineDispatcher
+    ) {
         val scope: CoroutineScope = CoroutineScope(dispatcher + job)
-    )
+        val runningJobs: AtomicLong = AtomicLong(0)
+    }
 
-    private val internalScope = ControlScope(Job(), Dispatchers.IO)
+    private val internalScope = ControlScope("internal", Job(), Dispatchers.IO)
 
     private val implementorScope =
         ControlScope(
-            SupervisorJob(),
+            "implementor",
+            Job(),
             Executors.newFixedThreadPool(config.maxNumImplementorTaskThreads)
                 .asCoroutineDispatcher()
         )
 
     override suspend fun launch(task: WrappedTask<ScopedTask>) {
-        when (task.innerTask) {
-            is InternalScope -> internalScope.scope.launch { execute(task, "internal") }
-            is ImplementorScope -> implementorScope.scope.launch { execute(task, "implementor") }
-            is ShutdownScope -> implementorScope.scope.launch { execute(task, "shutdown") }
+        val scope =
+            when (task.innerTask) {
+                is InternalScope -> internalScope
+                is ImplementorScope -> implementorScope
+            }
+        scope.scope.launch {
+            var nJobs = scope.runningJobs.incrementAndGet()
+            log.info { "Launching task $task in scope ${scope.name} ($nJobs now running)" }
+            val elapsed = measureTimeMillis { task.execute() }
+            nJobs = scope.runningJobs.decrementAndGet()
+            log.info { "Task $task completed in $elapsed ms ($nJobs now running)" }
         }
     }
 
-    private suspend fun execute(task: WrappedTask<ScopedTask>, scope: String) {
-        log.info { "Launching task $task in scope $scope" }
-        val elapsed = measureTimeMillis { task.execute() }
-        log.info { "Task $task completed in $elapsed ms" }
-    }
-
+    @OptIn(InternalCoroutinesApi::class)
     override suspend fun close() {
-        log.info { "Closing task scopes" }
         // Under normal operation, all tasks should be complete
         // (except things like force flush, which loop). So
         // - it's safe to force cancel the internal tasks
         // - implementor scope should join immediately
+        log.info { "Closing task scopes (${implementorScope.runningJobs.get()} remaining)" }
```

Comment:
is this basically equivalent? (just trying to avoid the internal api call)
```kotlin
        val throwable: AtomicReference<Throwable> = AtomicReference()
        implementorScope.job.children.forEach {
            it.invokeOnCompletion { cause ->
                if (cause != null) {
                    throwable.set(cause)
                }
            }
        }
        implementorScope.job.complete()
        implementorScope.job.join()
        if (throwable.get() != null) {
            throw IllegalStateException("Implementor job killed by uncaught exception", throwable.get())
        }
```

Reply:
That works against your branch, I'll add it here.